File: algorithms/238.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    vector<int> productExceptSelf(vector<int>& nums)
    {
        int length = nums.size();
        vector<int> result(length, 1);
        for ( int i = 1, tmp = 1; i < length; i++ )
        {
            tmp *= nums[i-1];
            result[i] *= tmp;
        }
        for ( int i = length-2, tmp = 1; i >= 0; i-- )
        {
            tmp *= nums[i+1];
            result[i] *= tmp;
        }
        return result;
    }
};

int main()
{
    return 0;
}

File: algorithms/python/238.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        rls = [1] * n

        t = 1
        for i in range(1, n):
            t *= nums[i-1]
            rls[i] *= t

        t = 1
        for i in range(n-2, -1, -1):
            t *= nums[i+1]
            rls[i] *= t

        return rls
